riscv/src/continuations/bootloader.rs

```diff
@@ -32,6 +37,225 @@ pub const MEMORY_HASH_START_INDEX: usize = 2 * REGISTER_NAMES.len();
 
+/// This trait provides all the field specific types and implementations that the bootloader needs.
+///
+/// For now, this trait is implemented directly by each `FieldElement` type, because the hash functions (i.e., poseidon) are field-specific.
+pub(crate) trait BootloaderImpl<F> {
+    // number of field elements to represent a single 32-bit word.
+    const FE_PER_WORD: usize;
+    // type of a memory page (must be a FE array of a specific size, but we can't use associated constants to define it...)
+    type Page;
+    // type of a hash value (same as Page, a FE array of a specific size)
+    type Hash;
+    fn update_page(page: &mut Self::Page, idx: usize, word: u32);
+    fn hash_page(page: &Self::Page) -> Self::Hash;
+    fn hash_two(a: &Self::Hash, b: &Self::Hash) -> Self::Hash;
+    fn zero_page() -> Self::Page;
+    // iterate over a hash value as machine words as field elements (i.e., WORDS_PER_HASH * Self::FE_PER_WORD field elements),
+    fn iter_hash_as_fe(h: &Self::Hash) -> impl Iterator<Item = F>;
+    // iterate over the page words, in their field element representation
+    fn iter_page_as_fe(p: &Self::Page) -> impl Iterator<Item = F>;
+    // iterate over a word value in its field element representation
+    fn iter_word_as_fe(w: u32) -> impl Iterator<Item = F>;
+}
+
+pub(crate) struct BootloaderInputs<F: FieldElement, B: BootloaderImpl<F>> {
+    inputs: Vec<F>,
+    phantom: PhantomData<B>,
+}
+
+impl<F: FieldElement, B: BootloaderImpl<F>> BootloaderInputs<F, B> {
+    pub fn new<I: ExactSizeIterator<Item = u32>>(
+        register_values: Vec<F>,
+        merkle_tree: &MerkleTree<F, B>,
+        accessed_pages: I,
+    ) -> Self {
+        // initial register values
+        let mut inputs = register_values;
+        // final register values
+        inputs.extend_from_within(..);
+        let root_hash = merkle_tree.root_hash();
+        // initial hash
+        inputs.extend(B::iter_hash_as_fe(root_hash));
+        // final hash
+        inputs.extend(B::iter_hash_as_fe(root_hash));
+        // number of pages
+        inputs.extend(B::iter_word_as_fe(accessed_pages.len() as u32));
+        // page data
+        for page_index in accessed_pages {
+            let (page_data, page_hash, proof) = merkle_tree.get(page_index as usize);
+            inputs.extend(B::iter_word_as_fe(page_index));
+            inputs.extend(B::iter_page_as_fe(page_data));
+            inputs.extend(B::iter_hash_as_fe(page_hash));
+            for sibling in proof {
+                inputs.extend(B::iter_hash_as_fe(sibling));
+            }
+        }
+        BootloaderInputs {
+            inputs,
+            phantom: PhantomData,
+        }
+    }
+
+    /// The bootloader input that is equivalent to not using a bootloader, i.e.:
+    /// - No pages are initialized
+    /// - All registers are set to 0 (including the PC, which causes the bootloader to do nothing)
+    /// - The state at the end of the execution is the same as the beginning
```

Comment:
why is the state the same?

Reply:
i don't know, did not dig into this yet, just kept the bootloader behaviour as it was, but adapting for the small_field memory layout

Reply:
so these comments were there before already?